src/Sarif/Visitors/ReformattingVisitor.cs

```diff
@@ -43,23 +45,129 @@ public override Run VisitRun(Run node)
                 }
             }
 
+            if (node.Results != null)
+            {
+                var results = new List<Result>();
+
+                foreach (Result result in node.Results)
+                {
+                    results.Add(VisitResult(result));
+                }
+
+                node.Results = results;
+            }
+
             return node;
         }
 
-        internal FileData VisitDictionaryValueNullChecked(string key, FileData node)
+        public override PhysicalLocation VisitPhysicalLocation(PhysicalLocation node)
         {
-            Uri uri;
+            // If we have a binary region, there is no work to do
+            if (node.Region == null ||
+                (node.Region.ByteOffset > 0 || node.Region.ByteLength > 0))
+            {
+                goto Exit;
+            }
+
+            // If we already have a snippet, let's assume it is correct
+            if (node.Region.Snippet != null)
+            {
+                goto Exit;
+            }
 
-            if (!Uri.TryCreate(key, UriKind.RelativeOrAbsolute, out uri))
+            // If the user hasn't specified code snippet insertion, there is no work to do
+            bool workToDo = false;
+            workToDo |= node.Region != null && _dataToInsert.Includes(OptionallyEmittedData.CodeSnippets);
+
+            if (!workToDo)
+            {
+                goto Exit;
+            }
+
+            // If we have an absolute URL specified as a file that 
+            // exist on disk, there is work to do
+            if (node.FileLocation.Uri.IsAbsoluteUri &&
+                node.FileLocation.Uri.LocalPath != null &&
+                File.Exists(node.FileLocation.Uri.LocalPath))
+            {
+                Region region = node.Region;
+                string fileText = GetFileText(node.FileLocation.Uri);
+                GetCompleteFirstLineAssociatedWithRegion(region, fileText);
+            }
+
+            Exit:
+            return base.VisitPhysicalLocation(node);
+        }
+
+        internal static void GetCompleteFirstLineAssociatedWithRegion(Region region, string fileText)
+        {
+            if (region.IsBinaryRegion)
```

Comment:
Shouldn't need to test this because you tested it on Line 67.

Reply:
good eye, yes, but i want this test in the helper, so that it is complete. in product usage, i test for binary regions early as an optimization.

---
In reply to: [202102249](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202102249) [](ancestors = 202102249)